notebooks/03 Data cleaning.ipynb

```diff
@@ -0,0 +1,735 @@
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# fill IS_major\n",
+    "imputed_df.IS_major.fillna(value=\"N\", inplace=True)\n",
+    "imputed_df.IS_major.value_counts()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "## Drop sparse rows"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1650448285559
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# check for null values across dataset\n",
+    "imputed_df.isnull().sum()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1650448611763
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# some columns have < 1000 null values, and a larger subset have ~68000 missing. Drop these\n",
+    "removed_sparse_rows_df = imputed_df.dropna(\n",
+    "    subset=[\n",
+    "        \"ADMISSION_METHOD_HOSPITAL_PROVIDER_SPELL_DESCRIPTION\",\n",
+    "        \"EXPECTED_DISCHARGE_DATE\",\n",
+    "        \"MAIN_SPECIALTY_CODE_AT_ADMISSION_DESCRIPTION\",\n",
+    "        \"POST_CODE_AT_ADMISSION_DATE_DISTRICT\",\n",
+    "        \"IMD county decile\",\n",
+    "        \"all_diagnoses\",\n",
+    "    ]\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Remove duplicates"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "no_duplicate_rows_df = removed_sparse_rows_df.drop_duplicates()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "## Homogenise binary fields\n",
+    "\n",
+    "Many fields are encoding as Y/N or similar, convert these into binary fields"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1650449146256
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "binary_fields_df = no_duplicate_rows_df.copy()\n",
+    "binary_fields_df.stroke_ward_stay = binary_fields_df.stroke_ward_stay.apply(\n",
+    "    lambda x: 0 if x == \"N\" else 1\n",
```

Comment:
I wondered if numpy.where() might be more readable than using Lambda functions here?